### PR TITLE
Dispatch global checkpoint sync to write pool

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently the global checkpoint sync is place on the management thread
pool. In scenarios where the node is IO limited this can lead to task
build-up and saturate the management threadpool interfering with other
cluster operations. This commit dispatches these to the write thread
pool similar to write operations (write thread pool for normal request,
system write thread pool for system indices).
    
Additionally, it adds logic to only allow a single global checkpoint
sync to be scheduled at a time.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java [hunk 1]
```diff
@@ -77,18 +94,60 @@ protected void shardOperationOnPrimary(
...
+        IndexShard shard,
+        Supplier<R> resultSupplier,
+        ActionListener<R> listener
+    ) {
+        String executor = executorProvider.apply(shard);
+
+        PendingSyncs<R> pending = pendingSyncs.compute(shard, (ignored, existing) -> {
+            final PendingSyncs<R> toReturn;
+            toReturn = Objects.requireNonNullElseGet(existing, () -> new PendingSyncs<>(request));
+            toReturn.toComplete.add(new Tuple<>(resultSupplier, listener));
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java:117]: This works for `ConcurrentHashMap`, but not other maps, might be worth adding a comment to clarify.

**Review by DaveCTurner — COMMENTED** [review]
I think I'd prefer more of this functionality to be in `IndexShard` directly, rather than handled by maps keyed by `IndexShard` as done here.

Also I think we could use a `SingleResultDeduplicator` to do the deduplication, unless there's a difference in semantics that I'm missing here.

File: server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java [hunk 2]
```diff
@@ -77,18 +94,60 @@ protected void shardOperationOnPrimary(
...
+                protected void doRun() {
+                    PendingSyncs<R> removed = pendingSyncs.remove(shard);
+                    assert removed == pending;
+
+                    ConcurrentLinkedQueue<Tuple<Supplier<R>, ActionListener<R>>> toComplete = removed.toComplete;
+                    try {
+                        maybeSyncTranslog(shard);
+                        for (Tuple<Supplier<R>, ActionListener<R>> tuple : toComplete) {
+                            ActionListener.completeWith(tuple.v2(), () -> tuple.v1().get());
+                        }
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java:144]: Can we move the completion here outside the exception catch block? I think that makes it somewhat clearer, I would expect no exceptions from this.

**Review by henningandersen — COMMENTED** [review]
This looks good on it's own, though I wonder if some of the work here will be partially unnecessary if we were to capture the relevant seq-no when dispacthing the gcp-sync action and turning those into no-ops.

But I'd be fine to see this go in as is and then tackle the other part separately. So please go ahead and add tests.

**Comment by DaveCTurner** [comment]
On reflection I'm not sure that the deduplication mechanism proposed here will help so much once the work moves to the `WRITE` threads, because then the IO work to sync the GCPs will interleave with, and therefore effectively throttle, the actual write work. Maybe it'll help in some cases but I think there'll still be some indexing workloads that do too much GCP syncing.

In any case here's a sketch of what I meant about moving this functionality into `IndexShard`:

```diff
diff --git a/server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java b/server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java
index d466d0988ab..800f3ff463c 100644
--- a/server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java
+++ b/server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java
@@ -21,7 +21,6 @@ import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.ShardId;
-import org.elasticsearch.index.translog.Translog;
 import org.elasticsearch.indices.IndicesService;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.TransportService;
@@ -77,25 +76,12 @@ public class GlobalCheckpointSyncAction extends TransportReplicationAction<
         IndexShard indexShard,
         ActionListener<PrimaryResult<Request, ReplicationResponse>> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
-            maybeSyncTranslog(indexShard);
-            return new PrimaryResult<>(request, new ReplicationResponse());
-        });
+        indexShard.syncTranslogOnStaleGlobalCheckpoint(listener.map(ignored -> new PrimaryResult<>(request, new ReplicationResponse())));
     }
 
     @Override
-    protected void shardOperationOnReplica(Request shardRequest, IndexShard replica, ActionListener<ReplicaResult> listener) {
-        ActionListener.completeWith(listener, () -> {
-            maybeSyncTranslog(replica);
-            return new ReplicaResult();
-        });
-    }
-
-    private static void maybeSyncTranslog(final IndexShard indexShard) throws IOException {
-        if (indexShard.getTranslogDurability() == Translog.Durability.REQUEST
-            && indexShard.getLastSyncedGlobalCheckpoint() < indexShard.getLastKnownGlobalCheckpoint()) {
-            indexShard.sync();
-        }
+    protected void shardOperationOnReplica(Request shardRequest, IndexShard indexShard, ActionListener<ReplicaResult> listener) {
+        indexShard.syncTranslogOnStaleGlobalCheckpoint(listener.map(ignored -> new ReplicaResult()));
     }
 
     public static final class Request extends ReplicationRequest<Request> {
diff --git a/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java b/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
index 8392bb118e3..ebedf9368c1 100644
--- a/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
+++ b/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
@@ -30,6 +30,7 @@ import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.ActionRunnable;
+import org.elasticsearch.action.SingleResultDeduplicator;
 import org.elasticsearch.action.admin.indices.flush.FlushRequest;
 import org.elasticsearch.action.admin.indices.forcemerge.ForceMergeRequest;
 import org.elasticsearch.action.support.PlainActionFuture;
@@ -232,6 +233,10 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
 
     private final IndexingOperationListener indexingOperationListeners;
     private final Runnable globalCheckpointSyncer;
+    private final SingleResultDeduplicator<Void> globalCheckpointSyncDeduplicator = new SingleResultDeduplicator<>(
+        null,
+        this::innerSyncTranslogOnStaleGlobalCheckpoint
+    );
 
     Runnable getGlobalCheckpointSyncer() {
         return globalCheckpointSyncer;
@@ -3622,6 +3627,23 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
         return getEngine().isTranslogSyncNeeded();
     }
 
+    /**
+     * Sync translog if the last-synced global checkpoint is out of date.
+     */
+    public void syncTranslogOnStaleGlobalCheckpoint(ActionListener<Void> listener) {
+        globalCheckpointSyncDeduplicator.execute(listener);
+    }
+
+    private void innerSyncTranslogOnStaleGlobalCheckpoint(ActionListener<Void> listener) {
+        ActionListener.completeWith(listener, () -> {
+            if (getTranslogDurability() == Translog.Durability.REQUEST
+                && getLastSyncedGlobalCheckpoint() < getLastKnownGlobalCheckpoint()) {
+                sync();
+            }
+            return null;
+        });
+    }
+
     /**
      * Returns the current translog durability mode
      */
```

**Comment by Tim-Brooks** [comment]
Closing as we merged https://github.com/elastic/elasticsearch/pull/96364. I might look at bringing back some of the queue deduplication pieces from this. But that would be a different PR.

**Pull request closed without merging**